### PR TITLE
feat(worker): add MessageEvent, ErrorEvent and handling to Worker API

### PR DESCRIPTION
**Pull request by crowlKats**
@bartlomieju

### CODE REVIEW

File: cli/js/web/workers.ts [hunk 1]
```diff
@@ -28,6 +28,72 @@ function decodeMessage(dataIntArray: Uint8Array): any {
...
+  postMessage(message: any, transfer: object[]): void;
+  postMessage(message: any, options: PostMessageOptions): void;
+}
+export type MessageEventSource = MessagePort | Worker; // | WindowProxy
+export interface MessageEventInit extends __domTypes.EventInit {
+  data?: any;
+  origin?: string;
+  lastEventId?: string;
+  source?: MessageEventSource | null;
+  ports?: MessagePort[];
```
Comment (bartlomieju) [inline cli/js/web/workers.ts]: We don't support ports right now, it should throw error when `ports` option is passed

File: cli/js/web/workers.ts [hunk 2]
```diff
@@ -133,6 +201,11 @@ export class WorkerImpl extends EventTarget implements Worker {
           const message = decodeMessage(new Uint8Array(event.data));
           this.onmessage({ data: message });
         }
+        // TODO: this is being handled in a type unsafe way, it should be type safe
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const ev = new Event("message", { cancelable: false }) as any;
```
Comment (bartlomieju) [inline cli/js/web/workers.ts]: create `new MessageEvent`

File: cli/js/web/workers.ts [hunk 3]
```diff
@@ -109,9 +175,11 @@ export class WorkerImpl extends EventTarget implements Worker {
...
     if (this.onerror) {
       this.onerror(event);
-      if (event.defaultPrevented) {
-        handled = true;
-      }
+    }
+
+    this.dispatchEvent(event);
+    if (event.defaultPrevented) {
+      handled = true;
```
Comment (bartlomieju) [inline cli/js/web/workers.ts]: create `new ErrorEvent`

**Comment by crowlKats** [comment]
@bartlomieju updated

File: cli/js/web/workers.ts [hunk 4]
```diff
@@ -28,6 +29,75 @@ function decodeMessage(dataIntArray: Uint8Array): any {
...
+    if (eventInitDict.ports) {
+      notImplemented();
+    }
+
+    this.data = eventInitDict.data;
+    this.origin = eventInitDict.origin;
+    this.lastEventId = eventInitDict.lastEventId;
+    this.source = eventInitDict.source;
+  }
+}
```
Comment (nayeemrmn) [inline cli/js/web/workers.ts]: You're dereferencing `eventInitDict` which could be undefined. Use optional chaining, replace all the `.`s with `?.`s.

File: cli/js/web/workers.ts [hunk 5]
```diff
@@ -28,6 +29,75 @@ function decodeMessage(dataIntArray: Uint8Array): any {
...
+export class ErrorEvent extends Event {
+  readonly message: string;
+  readonly filename: string;
+  readonly lineno: number;
+  readonly colno: number;
+  readonly error: any;
+
+  constructor(type: string, eventInitDict?: ErrorEventInit) {
+    super(type, {
+      bubbles: eventInitDict.bubbles,
```
Comment (nayeemrmn) [inline cli/js/web/workers.ts]: ditto

File: cli/js/web/workers.ts (hunk 4 above)
Comment (nayeemrmn) [reply cli/js/web/workers.ts]: Mind the types of the fields. Either they need to be nullable now or you need to specify a default like shown above.

Comment (crowlKats) [reply cli/js/web/workers.ts]: whoops

**Comment by crowlKats** [comment]
@bartlomieju the test cases seem to fail

File: cli/js/web/workers.ts [hunk 6]
```diff
@@ -133,6 +205,12 @@ export class WorkerImpl extends EventTarget implements Worker {
           const message = decodeMessage(new Uint8Array(event.data));
           this.onmessage({ data: message });
```
Comment (bartlomieju) [inline cli/js/web/workers.ts]: MessageEvent should be passed as an argument to this callback as well

Comment (crowlKats) [reply cli/js/web/workers.ts]: sorry for the long wait, so like 
```suggestion
          this.onmessage(new MessageEvent("message", {data: message}));
```
?
should i change the type of `onmessage` param as well?

Comment (bartlomieju) [reply cli/js/web/workers.ts]: Yes, bit create only a single `new MessageEvent`

Comment (crowlKats) [reply cli/js/web/workers.ts]: as in
```ts
if (type === "msg") {
  const message = decodeMessage(new Uint8Array(event.data));
  const ev = new MessageEvent("message", {
    cancelable: false,
    data: this.onmessage ? message : event.data
  });

  this.dispatchEvent(ev);
  continue;
}
```
?

**Comment by aduh95** [comment]
Fixes #4344

**Comment by bartlomieju** [comment]
@crowlKats sorry for slow review, I merged `master` and did update your PR a bit. Good work 👍

**Pull request merged**